include/RAJA/internal/Span.hpp

```diff
@@ -60,26 +61,50 @@ struct Span {
   static_assert(type_traits::is_random_access_iterator<ValueType>::value,
                 "ValueType must model RandomAccessIterator");
 
-  iterator begin() { return _data; }
-  iterator end() { return _data + _length; }
-  const_iterator begin() const { return _data; }
-  const_iterator end() const { return _data + _length; }
-  const_iterator cbegin() const { return _data; }
-  const_iterator cend() const { return _data + _length; }
-
-  ValueType data() const { return _data; }
-  IndexType size() const { return _length; }
-  IndexType max_size() const { return _length; }
-  bool empty() const { return _length == 0; }
-
-  iterator _data;
-  IndexType _length;
+  RAJA_HOST_DEVICE Span(iterator begin, iterator end)
+    : m_begin{begin}, m_end{end}
+  {}
+
+  RAJA_HOST_DEVICE Span(iterator begin, IndexType size)
+    : m_begin{begin}, m_end{begin + size}
+  {}
+
+  RAJA_HOST_DEVICE RAJA_INLINE iterator begin() { return m_begin; }
+  RAJA_HOST_DEVICE RAJA_INLINE iterator end() { return m_end; }
+  RAJA_HOST_DEVICE RAJA_INLINE const_iterator begin() const { return m_begin; }
+  RAJA_HOST_DEVICE RAJA_INLINE const_iterator end() const { return m_end; }
+  RAJA_HOST_DEVICE RAJA_INLINE const_iterator cbegin() const { return m_begin; }
+  RAJA_HOST_DEVICE RAJA_INLINE const_iterator cend() const { return m_end; }
+
+  RAJA_HOST_DEVICE RAJA_INLINE ValueType data() const { return m_begin; }
+  RAJA_HOST_DEVICE RAJA_INLINE IndexType size() const { return static_cast<IndexType>(m_end - m_begin);}
+  RAJA_HOST_DEVICE RAJA_INLINE IndexType max_size() const { return static_cast<IndexType>(m_end - m_begin);}
+  RAJA_HOST_DEVICE RAJA_INLINE bool empty() const { return static_cast<IndexType>(m_end - m_begin) == 0; }
+
+  //returns a span wrapper starting at begin with length ``length``
+  RAJA_HOST_DEVICE RAJA_INLINE
+  Span slice(size_type begin,size_type length) const
+  {
+    auto start = m_begin + begin;
+    auto end = start + length > m_end ? m_end : start + length;
+    return Span(start,end);
+  }
+  iterator m_begin;
+  iterator m_end;
 };
 
 template <typename ValueType, typename IndexType>
-Span<ValueType, IndexType> make_span(ValueType data, IndexType size)
-{
-  return {data, size};
+RAJA_HOST_DEVICE RAJA_INLINE
+Span<ValueType, IndexType> make_span(ValueType begin, IndexType size)
+{  
+  return Span<ValueType,IndexType>(begin, begin + size);
+}
+
+template <typename ValueType, typename IndexType>
+RAJA_HOST_DEVICE RAJA_INLINE
+Span<ValueType, IndexType> make_span(ValueType begin, ValueType end)
```

Comment:
This maker function doesn't deduce IndexType.
It should be made to, or removed.